api/services/contributionService.ts

```diff
@@ -0,0 +1,135 @@
+            contribution.status = contributionAttrs.status;
+
+            contribution.contrPrefix = contributionAttrs.prefix;
+            contribution.contrFirst = contributionAttrs.firstName;
+            contribution.contrMiddleInitial = contributionAttrs.middleInitial;
+            contribution.contrLast = contributionAttrs.lastName;
+            contribution.contrSuffix = contributionAttrs.suffix;
+            contribution.contrTitle = contributionAttrs.title;
+            contribution.email = contributionAttrs.email;
+            contribution.address1 = contributionAttrs.address1;
+            contribution.address2 = contributionAttrs.address2;
+            contribution.city = contributionAttrs.city;
+            contribution.state = contributionAttrs.state;
+            contribution.zip = contributionAttrs.zip;
+            contribution.contrName = contributionAttrs.name;
+            contribution.contributorType = contributionAttrs.contributorType;
+
+            contribution.amount = contributionAttrs.amount;
+            contribution.matchAmount = contributionAttrs.matchAmount;
+            contribution.submitForMatch = contributionAttrs.submitForMatch ? contributionAttrs.submitForMatch : false;
+            if (await contribution.isValidAsync()) {
+                await contributionRepository.save(contribution);
+                return contribution;
+            }
+            throw new Error('Contribution is missing one or more required properties.');
+        }
+        throw new Error('User is not permitted to add contributions for this campaign.');
+    } catch (e) {
+        throw new Error(e.message);
+    }
+}
+
+export interface IGetContributionOptions {
+    currentUserId?: number;
+    campaignId?: number;
+    perPage?: number;
+    page?: number;
+    status?: string;
+    from?: string;
+    to?: string;
+}
+
+export interface IGetContributionAttrs extends IGetContributionOptions {
+    governmentId: number;
+}
+
+export async function getContributionsAsync(contributionAttrs: IGetContributionAttrs): Promise<IContributionSummary[]> {
+    try {
+        const { governmentId, ...options } = contributionAttrs;
+        if (options.campaignId) {
+            const hasCampaignPermissions =
+                (await isCampaignAdminAsync(options.currentUserId, options.campaignId)) ||
+                (await isCampaignStaffAsync(options.currentUserId, options.campaignId)) ||
+                (await isGovernmentAdminAsync(options.currentUserId, governmentId));
+            if (hasCampaignPermissions) {
+                return getContributionsByGovernmentIdAsync(governmentId, {
+                    ...options,
+                    page: options.page || 0,
+                    perPage: options.perPage || 100
+                });
+            }
+            throw new Error('User is not permitted to get contributions for this campaign.');
+        }
+        return getContributionsByGovernmentIdAsync(governmentId, {
```

Comment:
I think we need an else block here. If the user is a campaign admin or staff, they can see only their campaign contributions. If campaignId is not present, then it must be a government admin, but, we should check to make sure. Reading the code, I think i could see all campaign contributions for a government.

```        
const { governmentId, ...options } = contributionAttrs;
if (options.campaignId) {

} else {
if (!isGovernmentAdminAsync(options.currentUserId, governmentId)) {
throw new Error('Must be a government admin to see all contributions')}
}
```

Reply:
Resolved by commit 738e4d0